reactor/reactor-core
# Migrate reference documentation to Antora

This PR attempts to migrate Reactor Core reference documentation to Antora, and includes the following logical commits:

- a482fbbd007eb2f96fe1748e355acb40ae5f6f63: Moved asciidoc file paths so they are compliant with Antora structure:
- 2d340c040cb21309c3174d115a320d8d244fd6f6: Refactor ascidoc adoc files to Antora. core features and advenced features are now split in multiple adoc files
- 90e86ee64cc884bb0aec5ca71515cea1b3c0f61d: Add antora yml file descritors
- 33be68a3b15becb7598a61a80dde79d361c9eca0: Adapted gradle scripts for Antora
- eaa7bb44303843d8e1f25d06ed69c88037b004cb: Adapted the `publish` CI in order to configure JDK21 as the default JDK, because Antora gradle plugin requires a JDK17 compatible version (so far, I could not use toolchain to selectively run only the antora gradle plugin with JDK21, because Gradle loads all plugin classes in the current JVM used, see this [discussion](https://stackoverflow.com/questions/75625330/gradle-toolchain-resolver-plugin-doesnt-download-jdk17-for-a-spring-boot-3-proj) which might be related). Setting JDK21 as the default JDK should not affect generation of all jars because toolchains still ensures that jars are built with proper JDK version.  Note that if using JDK21 as the default JDK is really an issue, we can consider to do what was done for reactor-netty: use a separate job that builds only the doc in JDK21, then share the docs.zip between jobs using some upload/download GH actions ...). In the commit, the asciidoctor-pdf tool is also installed when releasing a RELEASE , or a MILESTONE version, because this tool is needed when converting antora to PDF (in snapshot mode, PDF is not generated, so asciidoctor-pdf tool is not installed in this case).
- acec348b730fb2df5030a32767b65ff560843b6c: Updated README to explain how to generate the documentation locally.

### To test this PR:

- Set the current JDK to JDK21 (or a JDK17 compatible version)
- generate the documentation:
```
./gradlew antora
``` 
you can browse docs/build/site/index.html

The search is not yet enabled, it will be enabled once we'll have merged this PR because it's needed to crawl the documentation using Algolia from **projectreactor.io** site.

### To test docs.zip generation:

- Set the current JDK to JDK21 (or a JDK17 compatible version)
- then type:
```
./gradlew docs
```

it will generate docs/build/distributions/reactor-core-3.7.0-SNAPSHOT-docs.zip

### To test publication:

- make sure all JDK versions are properly installed (see `Trouble building the project` section in the reactor core [README](https://github.com/reactor/reactor-core)).
- set current JDK version to JDK21.
- Run publish command:
```
./gradlew publishToMavenLocal
```
All jars should be built with proper jdk versions because toolchain is used.
- then check what is generated in your M2 repo

### To test publication in two steps:

If using JDK21 as the default JDK version is really an issue, you can then publish everything in two steps:

- step 1: Set current JDK to JDK21, and only run `./gradlew docs` : this will only build the docs.zip
- step 2: Set current JDK to JDK8, then run `./gradlew publishToMavenLocal`, this command will pick up the docs.zip previously built in step 1 and will include it in local M2

### To test PDF generation

As before, PDF are not generated by default in snapshot mode. If you want to generate a PDF, then do the following:

- first, you will need to install `asciidoctor-pdf` command, because the antora pdf assembler requires it:
```
brew install asciidoctor
```

- then generate antora documention + PDF:

```
./gradlew antora -PforcePdf
```

docs is generated as before in docs/build/site/index.html, but this time a PDF is also generated in ./docs/build/assembler/reactor/3.7.0/reactor-3-reference-guide.pdf

If you want to include the pdf in the docs.zip distribution file:

```
./gradlew docs -PforcePdf
```

and the pdf will be generated and included in the docs.zip:
```
unzip -l docs/build/distributions/reactor-core-3.7.0-SNAPSHOT-docs.zip |grep pdf
  4114787  05-03-2024 10:35   docs/reactor-core-reference-guide-3.7.0-SNAPSHOT.pdf
```

- you can force PDF generation also when publishing:
```
./gradlew publishToMavenLocal -PforcePdf
unzip -l ~/.m2/repository/io/projectreactor/reactor-core/3.7.0-SNAPSHOT/reactor-core-3.7.0-SNAPSHOT-docs.zip | grep pdf
  4114787  05-03-2024 10:37   docs/reactor-core-reference-guide-3.7.0-SNAPSHOT.pdf
```

as before, PDF is generated and included when releasing a RELEASE, or RC, or a MILESTONE, but not when releasing a SNAPSHOT version.

Fixes #3083
